### ISSUE TITLE
Enable install through pyproject.toml

### ISSUE THREAD

**Issue by matthewfeickert** [diana-hep/pyhf#399]
# Description

Given [PEP 518](https://www.python.org/dev/peps/pep-0518/) it would be nice to have the ability to install pyhf through the `pyproject.toml` (which already exists since PR #322) as an alternative (...or in the future a replacement) to using the `setup.py` with `pip`. This would also allow installation with dependency managers like [Poetry](https://poetry.eustace.io/).

**Comment by matthewfeickert** [comment]
Should also consider moving the bulk of `setup.py` to `setup.cfg` to take advantage of [`setuptools`'s work](https://github.com/pypa/setuptools/pull/1675) to support [PEP 517](https://www.python.org/dev/peps/pep-0517/)

c.f. [Bernat Gabor's post on this](https://www.bernat.tech/pep-517-518/)

**Comment by webknjaz** [comment]
Just drop in a `pyproject.toml` like this and that's it:
```toml
[build-system]
requires = [
  "setuptools >= 41.0.0",
  "setuptools_scm >= 1.15.0",
  "setuptools_scm_git_archive >= 1.0",
  "wheel",
]
build-backend = "setuptools.build_meta"
```

**Comment by matthewfeickert** [comment]
Given PR #455 this [Twitter thread](https://twitter.com/btskinn/status/1121633712209137665) from @bskinn will probably be useful in the future for doing this properly.

**Comment by webknjaz** [comment]
This mostly hit `tox` users (because it always uses the latest published pip): https://github.com/tox-dev/tox/issues/1276. But the maintainers are quite responsive and the fix should be out soon anyway :)

**Comment by matthewfeickert** [comment]
> This mostly hit tox users 

Yeah, it happened to us as well (PR #455). I've been watching [the Issue](https://github.com/pypa/pip/issues/6434), which should help in the future.

**Comment by webknjaz** [comment]
It's not the same: tox updates pip to the latest version and the workaround to this is quite non-trivial for tox specifically :)

**Comment by bskinn** [comment]
Huh. It must be a particular style of tox usage, because my tox (using it for factorial testing of different dependency matrices) works fine for two different projects.

**Comment by webknjaz** [comment]
You'll hit a bug only if you have a combo of `pyproject.toml` file existing in the project + editable mode.

**Comment by webknjaz** [comment]
The workaround, for now, is smth like this: https://github.com/ansible/molecule/pull/2009

**Comment by bskinn** [comment]
Interesting... what's the use case for an editable install inside a tox environment? Post-install inspection/meddling?

**Comment by webknjaz** [comment]
Just for development purposes. By default, tox doesn't reinstall the env. Unless you make it.
And `use_develop = True` is it's default setting.
This way, when you rerun tests they are executed against existing env + your source code is already linked there. Of course, this assumes you don't have something that has to be compiled every time.
